noise_corrector.py

```diff
@@ -1,70 +1,177 @@
+  assert len(observed_values) == 3
+  randomized_response_rate = .0000025
+  return estimate_true_values(observed_values, randomized_response_rate)
+
+
+def correct_nav_sources(observed_values):
+  assert len(observed_values) == 2925
+  randomized_response_rate = .0024
+  return estimate_true_values(observed_values, randomized_response_rate)
+
+
+def simulate_randomization(true_reports, randomized_response_rate):
+  """Simulates randomized response on true_reports"""
+  n = sum(true_reports)
+  k = len(true_reports)
+  noisy_reports = [0] * k
+
+  # This method could be implemented by doing randomized response on every
+  # individual true report. However, we can optimize this by noticing that
+  # every bucket's final result will be distributed according to a Binomial
+  # distribution with parameter n and q = (1 - x)*(v' / n) + xr
+  # where x = randomized_response_rate. Note that this is just an estimate,
+  # since the true randomized response will not be independent for each value.
+  x = randomized_response_rate
+
+  def gen_noisy(true_count):
+    q = (1 - x) * (true_count / n) + x / k
+    return numpy.random.binomial(n, q)
+  return [gen_noisy(c) for c in true_reports]
 
-  # Subtract the reports added to each bucket due to noise, and rescale to
-  # account for the reports that were shifted due to the initial noise.
-  corrected_buckets = {
-      bucket: (v - noised_values_per_bucket) / (1 - noise_probability)
-      for bucket, v in buckets.items()
-  }
-  return corrected_buckets
 
 if __name__ == "__main__":
-  # The following is an example showing how to use the function.
-  # |example_reports| is a map from bucket --> count of conversions
-  # with that bucket.
-  example_reports = {
-      0: 50,
-      1: 150,
-      2: 300,
-      3: 400,
-      4: 700,
-      5: 200,
-      6: 0,
-      7: 2000
-  }
-
-  # Simulate the API randomly flipping reports
-  noisy_reports = collections.defaultdict(int)
-  buckets = list(example_reports.keys())
-  noise_probability = .05
-  for bucket, count in example_reports.items():
-    for r in range(count):
-      if random.random() <= noise_probability:
-        new_bucket = random.choice(buckets)
-        noisy_reports[new_bucket] += 1
-      else:
-        noisy_reports[bucket] += 1
-
-  corrected = corrected_buckets(noisy_reports, noise_probability)
-
-  column_names = ["Bucket", "True count", "Noisy count", "Corrected count"]
-  print("{:<20}{:<20}{:<20}{:<20}".format(*column_names))
-  for bucket, count in sorted(corrected.items()):
-    print(f"{bucket:<20}{example_reports[bucket]:<20}{noisy_reports[bucket]:<20}{count:<20.2f}")
+  # This example runs sample data through a simulation of the randomized
+  # response, and applies the noise correction to it.
+  true_reports = [
+      100_000_000,  # Sources with no reports
+      100_000,  # Sources with conversion metadata "0"
+      40_000,  # Sources with conversion metadata "1"
+  ]
+  randomized_response_rate = .0000025
```

Comment:
This is fixed by using a global variable, which I did in the recent commit. It shouldn't be passed as an arg to `correct_event_sources` because that function should never use any other rate.

Reply:
Gotcha - the global at the top of the file is helpful